gensim/models/ensemblelda.py

```diff
@@ -0,0 +1,1268 @@
+        # or there is no cluster
+        if min_cores == 0:
+            min_cores = 1
+
+        if min_cores is None:
+            # min_cores is a number between 1 and 3, depending on the number of models
+            min_cores = min(3, max(1, int(self.num_models / 4 + 1)))
+
+        results = self.cluster_model.results
+
+        # first, group all the learned cores based on the label,
+        # which was assigned in the cluster_model. The result is a
+        # dict of {group: [topic, ...]}
+        grouped_by_labels = {}
+        for topic in results:
+            if topic["is_core"]:
+                topic = topic.copy()
+
+                # counts how many different labels a core has as parents
+                topic["amount_parent_labels"] = len(topic["parent_labels"])
+
+                label = topic["label"]
+                if label not in grouped_by_labels:
+                    grouped_by_labels[label] = []
+                grouped_by_labels[label].append(topic)
+
+        # then aggregate their amount_parent_labels by maxing and parent_labels by concatenating
+        # the result is sorted by amount_parent_labels and stored in sorted_clusters
+        sorted_clusters = []
+        for label, group in grouped_by_labels.items():
+            amount_parent_labels = 0
+            parent_labels = []  # will be a list of sets
+            for topic in group:
+                amount_parent_labels = max(topic["amount_parent_labels"], amount_parent_labels)
+                parent_labels.append(topic["parent_labels"])
+            # - here, nan means "not yet evaluated"
+            # - removing empty sets from parent_labels
+            sorted_clusters.append({
+                "amount_parent_labels": amount_parent_labels,
+                "parent_labels": [x for x in parent_labels if len(x) > 0],
+                "is_valid": np.nan,
+                "label": label
+            })
+
+        # start with the most significant core.
+        sorted_clusters = sorted(sorted_clusters,
+            key=lambda cluster: (
+                cluster["amount_parent_labels"],
+                cluster["label"]  # makes sorting deterministic
+            ), reverse=True)
+
+        def remove_from_all_sets(label):
+            """removes a label from every set in "parent_labels" for
+            each core in sorted_clusters."""
+            for a in sorted_clusters:
+                if label in a["parent_labels"]:
+                    a["parent_labels"].remove(label)
+
+        # APPLYING THE RULES 1 to 3
+        # iterate over the cluster labels, see which clusters/labels
+        # are valid to cause the creation of a stable topic
+        for i, cluster in enumerate(sorted_clusters):
+            label = cluster["label"]
+            # label is iterating over 0, 1, 3, ...
+
+            # 1. rule - remove if the cores in the cluster have no parents
+            if cluster["amount_parent_labels"] == 0:
```

Comment:
Instead of these multiple conditionals nested inside the for loop, create the following function:

```python
def _is_valid_cluster(cluster, min_cores):
    """Document your three rules inside this function."""
    if cluster["amount_parent_labels"] == 0:
        return False
    elif len(cluster["parent_labels"]) < min_cores:
        return False
    elif cluster["amount_parent_labels"] < 1:  # Seems to overlap with the first condition
        return False
    return sum(map(lambda x: x == {label}, cluster["parent_labels"])) >= min_cores
```

Then you could simplify your loop as:

```python
for i, cluster in enumerate(sorted_clusters):  # totally untested code
    if not _is_valid_cluster(cluster, num_cores):
        cluster.update(is_valid=False)
        remove_from_all_sets(label)
```

See how things are simpler and more readable this way?